File: CSCI_313-323_Data_Structures_And_Algorithms/CSCI_313_Data_Structures/ExCode/08-Stack/LeetCode/LeetCode-0496.cpp
class Solution {
public:
    /// An O(nxm) algorithm that uses O(m) extra space (monotonic stack)
    vector<int> nextGreaterElement(vector<int>& nums1, vector<int>& nums2) {
        stack<int> S;

        vector<int> nextGreater(nums2.size());
        for (int i=(int)nums2.size()-1; i>=0; i--){
            while (!S.empty() && nums2[i] >= S.top()) S.pop();

            if (S.empty()) nextGreater[i] = -1;
            else           nextGreater[i] = S.top();

            S.push(nums2[i]);
        } //end-for

        // Find the index of nums1[i] inside nums2, and push the nextGreater to the result
        vector<int> result;
        for (int i=0; i<nums1.size(); i++){
            for (int j=0; j<nums2.size(); j++){
                if (nums1[i] == nums2[j]){
                    result.push_back(nextGreater[j]);
                    break;
                } //end-if
            } //end-for
        } //end-for

        return result;
    } //end-nextGreaterElement
};

File: CSCI_313-323_Data_Structures_And_Algorithms/CSCI_313_Data_Structures/ExCode/08-Stack/LeetCode/LeetCode-0496.java
class Solution {
    /// An O(nxm) algorithm that uses O(m) extra space (monotonic stack)
    public int[] nextGreaterElement(int[] nums1, int[] nums2) {
        Stack<Integer> S = new Stack<>();

        int[] nextGreater = new int[nums2.length];
        for (int i=nums2.length-1; i>=0; i--){
            while (!S.isEmpty() && nums2[i] >= S.peek()) S.pop();

            if (S.isEmpty()) nextGreater[i] = -1;
            else             nextGreater[i] = S.peek();

            S.push(nums2[i]);
        } //end-for

        // Find the index of nums1[i] inside nums2, and push the nextGreater to the result
        int [] result = new int[nums1.length];
        for (int i=0; i<nums1.length; i++){
            for (int j=0; j<nums2.length; j++){
                if (nums1[i] == nums2[j]){
                    result[i] = nextGreater[j];
                    break;
                } //end-if
            } //end-for
        } //end-for

        return result;
    } //end-nextGreaterElement
}